File: worker/worker.py
import json
import os
from random import randint
from time import sleep

import redis

print('starting worker...')

# if __name__ == '__main__':
redis_host = os.getenv('REDIS_HOST', 'auth_queue')
r = redis.Redis(host=redis_host, port=6379, db=0)
print('connected to redis!')
while True:
    print('looking for messages..')
    sleep(3)
    try:
        print(r.blpop('register_message_queue'))
        # mensagem = json.loads(r.blpop('register_message_queue')[1])
        # # Simulando envio de email...
        # print("#################### Enviando mensagem")
        # sleep(randint(5, 10))
        # print('#################### Mensagem enviada com sucesso: ' + mensagem)
    except Exception as e:
        print('Erro!!!!')

File: worker/src/worker.js
/* setup */
const redis_mq = require("rsmq");
const rsmq = new redis_mq({ host: 'auth_queue', port: 6379, ns: 'rsmq' });
const transporter = require("nodemailer").createTransport({
    host: "smtp.gmail.com",
    port: 465,
    secure: true,
    auth: {
        user: "bt.auth.service@gmail.com",
        pass: "BTAuthService!"
    },
});



/* @TODO: why does ```while (true)``` breaks this? */
setInterval(() => {
    if (rsmq.connected) {
        rsmq.popMessage({ qname: "register_message_queue" }, async(error, data) => {
            if (error) {
                return
            }

            if (data.id) {
                const message = JSON.parse(data.message);
                console.log("Message received and deleted from queue", data);
                try {
                    await transporter.sendMail({
                        from: 'BT Auth Service',
                        to: message.email,
                        subject: "Welcome to the BT Auth Service",
                        text: `Welcome to my Authentication Microservice, ${message.name}!`,
                        html: `<h3>Welcome to my Authentication Microservice, ${message.name}!</h3>`,
                    });
                } catch (e) {
                    console.log('error sending message, it was re-added to the queue')
                    await rsmq.sendMessageAsync({ qname: 'register_message_queue', message: JSON.stringify({ email: message.email, name: message.name }) });
                }
            }
        });
    } else {
        console.log('could not scan queue! not connected to redis!');
    }
}, 1000);